### DISCUSSION TITLE
HxGrid bug - The type or namespace name 'GridHeaderCellContext' does not exist in the namespace

### DISCUSSION THREAD

**Discussion by oscar230** [havit/Havit.Blazor#620] [Q&A] [answered]
Hi i get this error when i try to edit a header in a `HxGrid `as described in [the docs](https://havit.blazor.eu/components/hxgrid#header-filtering). I usually do this, but now it does not work.

* Blazor Server on ASP NET 6.0 LTS
* `Havit.Blazor.Components.Web.Bootstrap` version 4.2.2

# Error:
```ps1
Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0234	The type or namespace name 'GridHeaderCellContext' does not exist in the namespace '__Blazor.Havit.Blazor.Components.Web.Bootstrap' (are you missing an assembly reference?)	PROJECTNAME	C:\Users\Administrator\Source\Repos\PROJECTNAME\PROJECTNAME\Microsoft.NET.Sdk.Razor.SourceGenerators\Microsoft.NET.Sdk.Razor.SourceGenerators.RazorSourceGenerator\Pages_Index_razor.g.cs	332	Active
```

# Explanation
I get the error whenever i try to use `<HeaderTemplate>` in a `<HxGrid>` something which i commonly do in other projects.
Example in `index.razor`
```csharp
@page "/"
@using System.Globalization

<HxGrid TItem="CultureInfo" Responsive="true" DataProvider="DiagnosticFilesGridComponentGetGridData" ContentNavigationMode="GridContentNavigationMode.Pagination">
    <Columns>
        <HxGridColumn HeaderText="File last Written" ItemTextSelector="@(item => item.DisplayName)" SortKeySelector="@(item => item.DisplayName)">
            <HeaderTemplate>
                <HxInputDate @bind-Value="filterDate" Placeholder="Filter by date" InputSize="InputSize.Small" />
            </HeaderTemplate>
        </HxGridColumn>
        <HxContextMenuGridColumn Context="item">
            <HxButtonGroup>
                <HxButton Color="ThemeColor.Primary" >View</HxButton>
            </HxButtonGroup>
        </HxContextMenuGridColumn>
    </Columns>
</HxGrid>

@code {
    [Parameter]
    public int? AgentIdParameter { get; set; }

    private string agentsGridComponentNameFilter = string.Empty;
    private string filterDiagnosticsFileName = string.Empty;
    private DateTime? filterDate;
    private bool agentsGridComponentNameSearchOnlyIsDead = false;

    private async Task<GridDataProviderResult<CultureInfo>> DiagnosticFilesGridComponentGetGridData(GridDataProviderRequest<CultureInfo> request)
    {
        return request.ApplyTo(new List<CultureInfo>());
    }
}
```

Project file
```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <Nullable>enable</Nullable>
    <ImplicitUsings>enable</ImplicitUsings>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Havit.Blazor.Components.Web" Version="4.2.2" />
    <PackageReference Include="Havit.Blazor.Components.Web.Bootstrap" Version="4.2.2" />
  </ItemGroup>

</Project>
```

I've disassembled my DLL `Havit.Blazor.Components.Web.Bootstrap.dll` and i cannot find the record `GridHeaderCellContext` in namespace `__Blazor.Havit.Blazor.Components.Web.Bootstrap` (first image) only in `Havit.Blazor.Components.Web.Bootstrap` (second image). I know this double underscore path is a compiled namespace, altough the record should still reside there i believe.
![image](https://github.com/havit/Havit.Blazor/assets/11545617/9c49238d-7a91-4ecc-97a1-9a996ff30524)
![image](https://github.com/havit/Havit.Blazor/assets/11545617/0b504316-30d7-4105-a5b8-e4bda5ab287e)

The full build output is:
```ps
Build started...
------ Build started: Project: PROJECTNAME, Configuration: Debug Any CPU ------
Restored C:\Users\Administrator\Source\Repos\PROJECTNAME\PROJECTNAME\PROJECTNAME.csproj (in 21 ms).
C:\Users\Administrator\Source\Repos\PROJECTNAME\PROJECTNAME\Microsoft.NET.Sdk.Razor.SourceGenerators\Microsoft.NET.Sdk.Razor.SourceGenerators.RazorSourceGenerator\Pages_Index_razor.g.cs(332,503,332,524): error CS0234: The type or namespace name 'GridHeaderCellContext' does not exist in the namespace '__Blazor.Havit.Blazor.Components.Web.Bootstrap' (are you missing an assembly reference?)
Done building project "PROJECTNAME.csproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

# Note
* The actual project name is redacted and replaced with `PROJECTNAME`. 
* I use the official nuget repository from MS.

**Comment by hakenr** [comment]
This is a common Blazor issue usually caused by naming collision or missing `@using`.

See related issues:
- https://github.com/dotnet/aspnetcore/issues/37482
- https://github.com/dotnet/aspnetcore/issues/18757

Or check the pages describing the symptom and possible causes:
https://www.google.com/search?q=CS0234+The+type+or+namespace+name+does+not+exist+in+the+namespace+__Blazor

(To double-check, I tried your repro-code in our `BlazorTestApp` project. Both `net7.0` and `net6.0` builds work like a charm.)

**Comment by oscar230** [comment, marked as answer]
Thanks for your answer @hakenr , **although last night i fixed it by restarting by VM and updating (by reinstalling) Visual Studio Enterprise 2022 to version 17.7.**

I changed nothing in the code before and after the update so I don't know about missing `@using` but thanks, I noticed that problem in other projects, its quite common sadly. Hopefully the compiler will producer better error messages in the future.

You can mark this as fixed if someone comes along a similar problem, hopefully this will be usefull for them.
_Thanks for a great library! 🎉😊
// Oscar_

**Comment by hakenr** [comment]
Hypothesis: It seems the VS reinstall updated your .NET SDK to newer version where there were some fixes in Razor compiler.

**Reply by oscar230** [reply]
Yes, that must have been it! 👍 
![ThumbsThumbsUpKidGIF](https://github.com/havit/Havit.Blazor/assets/11545617/33933a2a-1a84-469b-8f32-161b5ad4ca59)
